tmk_core/common/action_layer.c

```diff
@@ -110,94 +110,79 @@ void layer_debug(void)
 }
 #endif
 
-#if !defined(NO_ACTION_LAYER) && defined(PREVENT_STUCK_MODIFIERS)
-uint8_t source_layers_cache[(MATRIX_ROWS * MATRIX_COLS + 7) / 8][MAX_LAYER_BITS] = {0};
-
-void update_source_layers_cache(keypos_t key, uint8_t layer)
-{
-    const uint8_t key_number = key.col + (key.row * MATRIX_COLS);
-    const uint8_t storage_row = key_number / 8;
-    const uint8_t storage_bit = key_number % 8;
-
-    for (uint8_t bit_number = 0; bit_number < MAX_LAYER_BITS; bit_number++) {
-        source_layers_cache[storage_row][bit_number] ^=
-            (-((layer & (1U << bit_number)) != 0)
-             ^ source_layers_cache[storage_row][bit_number])
-            & (1U << storage_bit);
+#if !defined(NO_ACTION_LAYER) && defined(PREVENT_STUCK_KEYS)
+static uint8_t layers_cache[(MATRIX_ROWS * MATRIX_COLS + 7) / 8][MAX_LAYER_BITS];
+
+static void update_layers_cache(keypos_t key, uint8_t layer);
+static uint8_t read_layers_cache(keypos_t key);
+
+static void update_layers_cache(keypos_t key, uint8_t layer) {
+    uint8_t number = key.col + key.row * MATRIX_COLS;
+    uint8_t row = number / 8;
+    uint8_t mask = 1U << number % 8;
+    for (int8_t column = MAX_LAYER_BITS - 1; column >= 0; --column) {
+        mask_byte(&layers_cache[row][column], layer & _BV(column), mask);
     }
 }
 
-uint8_t read_source_layers_cache(keypos_t key)
-{
-    const uint8_t key_number = key.col + (key.row * MATRIX_COLS);
-    const uint8_t storage_row = key_number / 8;
-    const uint8_t storage_bit = key_number % 8;
+static uint8_t read_layers_cache(keypos_t key) {
+    uint8_t number = key.col + key.row * MATRIX_COLS;
+    uint8_t row = number / 8;
+    uint8_t mask = 1U << number % 8;
     uint8_t layer = 0;
-
-    for (uint8_t bit_number = 0; bit_number < MAX_LAYER_BITS; bit_number++) {
-        layer |=
-            ((source_layers_cache[storage_row][bit_number]
-              & (1U << storage_bit)) != 0)
-            << bit_number;
+    for (int8_t column = MAX_LAYER_BITS - 1; column >= 0; --column) {
+        if (layers_cache[row][column] & mask) layer |= _BV(column);
     }
-
     return layer;
 }
-#endif
 
-/*
- * Make sure the action triggered when the key is released is the same
- * one as the one triggered on press. It's important for the mod keys
- * when the layer is switched after the down event but before the up
- * event as they may get stuck otherwise.
- */
-action_t store_or_get_action(bool pressed, keypos_t key)
-{
-#if !defined(NO_ACTION_LAYER) && defined(PREVENT_STUCK_MODIFIERS)
-    if (disable_action_cache) {
-        return layer_switch_get_action(key);
+uint8_t find_layer(keypos_t key) {
```

Comment:
If one of those functions will need a change in the future, it would be easy to forget to update the other one too. From the point of view of maintenance, the common code should be shared as much as possible.



On 11 May 2016 14:32, at 14:32, Eric Tang <notifications@github.com> wrote:
>>  
>> -/*
>> - * Make sure the action triggered when the key is released is the
>same
>> - * one as the one triggered on press. It's important for the mod
>keys
>> - * when the layer is switched after the down event but before the up
>> - * event as they may get stuck otherwise.
>> - */
>> -action_t store_or_get_action(bool pressed, keypos_t key)
>> -{
>> -#if !defined(NO_ACTION_LAYER) && defined(PREVENT_STUCK_MODIFIERS)
>> -    if (disable_action_cache) {
>> -        return layer_switch_get_action(key);
>> +uint8_t find_layer(keypos_t key) {
>
>I thought it would be clearer to separate them. Since only one of these
>functions is used at a time, there is no harm.
>
>---
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Vifon/qmk_firmware/pull/4/files/a60a39bd985b7a467fb7b005916096d0ef2378d8#r62838921